File: Java/src/Medium/q199.java
package Medium;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;

/**
 * Leetcode 199: Binary Tree Right View Traversal
 * @author Viunec
 */
public class q199 {
    private static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public static List<Integer> rightSideView(TreeNode root) {
        if (root == null)
            return new ArrayList<>();

        List<Integer> res = new ArrayList<>();
        Queue<TreeNode> queue = new ArrayDeque<>();

        int count;
        List<Integer> currentLevel = new ArrayList<>();;
        TreeNode currentNode;

        queue.add(root);
        while (!queue.isEmpty()) {
            count = queue.size();

            for (int i=0;i<count;i++) {
                currentNode = queue.poll();
                currentLevel.add(currentNode.val);

                if (currentNode.left != null) queue.add(currentNode.left);
                if (currentNode.right != null) queue.add(currentNode.right);
            }

            res.add(currentLevel.get(currentLevel.size() - 1));
            currentLevel.clear();
        }

        return res;
    }

    public static void testSample() {
        TreeNode root = new TreeNode(1);
        TreeNode n1 = new TreeNode(2);
        TreeNode n2 = new TreeNode(3);
        TreeNode n3 = new TreeNode(5);
        TreeNode n4 = new TreeNode(4);

        n1.right = n3;
        n2.right = n4;
        root.left = n1;
        root.right = n2;

        System.out.println(rightSideView(root));
    }
}

File: Python/src/medium/q199.py
from typing import Optional, List
from collections import deque 

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    """
    Leetcode 199: Binary Tree Right Side View
    """

    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        self.views = []
        self.bfs(root)
        return self.views

    def bfs(self, root:Optional[TreeNode]):
        if not root:
            return
        
        numCurr, numNext = 1,0
        queue = deque([root])

        while queue:
            curr = queue.popleft()
            numCurr -= 1

            if curr.left:
                queue.append(curr.left)
                numNext += 1
            if curr.right:
                queue.append(curr.right)
                numNext += 1

            if numCurr == 0:
                # right-most node
                self.views.append(curr.val)
                numCurr = numNext
                numNext = 0